polysemy-plugin/src/Polysemy/Plugin/InlineRecursiveCalls.hs

```diff
@@ -1,91 +1,153 @@
-             (idType n)
-         $ setInlinePragInfo vanillaIdInfo neverInlinePragma
-  pure [ (lazySetIdInfo n info', replace n n' b)
-       , (n', Var n)
-       ]
-
-
--- TODO(sandy): Make this only break loops in functions whose type ends in `Sem
--- * * -> Sem * *` for wildcards `*`
-loopbreakBinds
-    :: Bind CoreBndr
-    -> CoreSupplyM (Bind CoreBndr)
-loopbreakBinds nr@(NonRec n b)
-  | containsName n b = Rec <$> loopbreaker n b
-  | otherwise        = pure nr
-loopbreakBinds (Rec bs) = fmap (Rec . join) . for bs $ \(n, b) ->
-  case containsName n b of
-    False -> pure [(n, b)]
-    True  -> loopbreaker n b
 
+import Bag
+import GhcPlugins hiding ((<>))
+import HsSyn
+
+
+------------------------------------------------------------------------------
+-- | Forces compiler to inline functions by creating loopbreaker with
+-- NO_INLINE pragma, changing recursive calls to use it and by adding INLINE
+-- pragma to the original function.
+inlineRecCallsAction :: (MonadUnique m, HasDynFlags m)
+                     => HsGroup GhcRn -> m (HsGroup GhcRn)
+inlineRecCallsAction group = do
+  dyn_flags <- getDynFlags
+
+  ( if optLevel dyn_flags > 0
+      then inlineRecCalls
+      else pure
+    ) group
+
+------------------------------------------------------------------------------
+inlineRecCalls :: MonadUnique m => HsGroup GhcRn -> m (HsGroup GhcRn)
+inlineRecCalls group@(hs_valds -> XValBindsLR (NValBinds binds sigs)) = do
+
+  (binds', extra_sigs) <- second concat . unzip
+                      <$> traverse inlineRecCall (pairTypes binds sigs)
+
+  pure group{ hs_valds = XValBindsLR $ NValBinds binds' $ sigs ++ extra_sigs }
+
+inlineRecCalls _ = error "inlineRecCalls: expected renamed group"
+
+------------------------------------------------------------------------------
+-- | Pairs single binds with their types if they have one specified.
+pairTypes :: [(RecFlag, LHsBinds GhcRn)]
+          -> [LSig GhcRn]
+          -> [((RecFlag, LHsBinds GhcRn), Maybe (LHsSigWcType GhcRn))]
+-- TODO: better way of doing this would be great
+pairTypes binds sigs = flip evalState sigs $ foldM go [] binds where
+  go pairs = \case
+    b@(_, bagToList -> [L _ FunBind{ fun_id = L _ fun_name }]) -> do
+      sig <- findMapPop $ \case
+        L _ (TypeSig _ sig_names bind_type)
+          | elem fun_name $ unLoc <$> sig_names -> Just bind_type
+        _                                       -> Nothing
+      pure $ (b, sig) : pairs
+
+    b -> pure $ (b, Nothing) : pairs
+
+------------------------------------------------------------------------------
+-- | Takes binding group and type of binding inside if there's only one and
+-- returns group with inlining of recursive calls and signatures to be added
+-- to the environment.
+inlineRecCall :: MonadUnique m
+              => ((RecFlag, LHsBinds GhcRn), Maybe (LHsSigWcType GhcRn))
+              -> m ((RecFlag, LHsBinds GhcRn), [LSig GhcRn])
+inlineRecCall (b, mtype)
+  | (Recursive, (bagToList -> [
```

Comment:
I find this guard super hard to read.

Reply:
Moved pattern matching on `FunBind` to guard.